### ISSUE TITLE
explore alternative impl for cycle handling

### ISSUE THREAD

**Issue by nikomatsakis** [salsa-rs/salsa#193]
In #147, I noted an alternative impl strategy for cycle handling that I wanted to explore:

> Anyway, I also started a branch that was implemented a bit differently. What I was trying to do was to have each active query have an associated `Arc` with a pointer to the next active query on the stack (i.e., what it is blocked on). This way you could follow a chain of queries forward across threads pretty cleanly and lazily. I seem to recall that the code in this PR was a bit more complex -- but then I don't think I've looked at it since you rebased. 

Filing this issue so I don't forget. Still hoping to explore this, but I'd also be happy to work with someone else on it.

**Comment by nikomatsakis** [comment]
Motivation is at least in part to [avoid instantiating a vector for cross-thread calls](https://github.com/salsa-rs/salsa/pull/147#pullrequestreview-290450757):

https://github.com/salsa-rs/salsa/blob/a9860bf37f08c2b8029d62aacca60ea43c35d767/src/runtime.rs#L716-L720

But also to simplify the handling of cross-thread cycle collection generally by having a more standard graph.

**Comment by carljm** [comment]
@nikomatsakis Is this still relevant after #603 ?

**Comment by Veykril** [comment]
I believe we can close this now, the motivating comment is no longer relevant either
